### PR TITLE
Replace black and isort with ruff

### PR DESCRIPTION
**Pull request by shchur**
*Issue #, if available:* #3353

*Description of changes:*
- This PR replaces all references to `black` and `isort` with the `ruff` linter.

*Left for a future PR:*
- Replacing flake8 with ruff

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: core/pyproject.toml [hunk 1]
```diff
@@ -1,2 +1,34 @@
...
+    "smart_open",
+    "lightning",
+    "torchmetrics",
+    "transformers",
+    "nptyping",
+    "omegaconf",
+    "pytorch-metric-learning",
+    "nlpaug",
+    "nltk",
+]
```
Comment (shchur) [inline core/pyproject.toml:34]: We need to duplicate the `tool.ruff.lint.isort` sections in all `pyproject.toml` files for the ruff import sorter to work correctly.

File: core/pyproject.toml [hunk 2]
```diff
@@ -1,2 +1,34 @@
-[tool.black]
+[tool.ruff]
```
Comment (prateekdesai04) [inline core/pyproject.toml:1]: @Innixma @shchur would this change detect all the src paths for all modules after pip install ?
as that was the problem earlier

**Review by prateekdesai04 — APPROVED** [review]
LGTM

File: core/pyproject.toml (hunk 2 above)
Comment (Innixma) [reply core/pyproject.toml:1]: Probably not. But we can check after. My current plan is to remove the `pyproject.toml` files except for the root one after we converge on 119 line-length.

**Review by Innixma — APPROVED** [review]
LGTM! Thanks for sending this one out!

**Comment by Innixma** [comment]
@shchur I didn't see any edits to the `check_style` tests, example: https://github.com/autogluon/autogluon/blob/master/tabular/tests/test_check_style.py

These tests run `flake8`. Maybe they should run `ruff` instead? Will they crash given that we are trying to remove `flake8` as a test time dependency?

File: pyproject.toml [hunk 3]
```diff
@@ -1,10 +1,14 @@
-[tool.black]
+[tool.ruff]
 line-length = 119
-target-version = ['py38', 'py39', 'py310']
-
-[tool.isort]
-known_first_party = "autogluon"
-known_third_party = [
+target-version = "py38"
```
[the remark this answers is not in the document]
Comment (shchur) [reply pyproject.toml:3]: Ruff requires specifying the lower bound ([docs](https://docs.astral.sh/ruff/settings/#target-version)), so `target-version = "py38"` means that all versions `>= 3.8` are supported.

**Comment by shchur** [comment]
@Innixma my plan is to deal with `flake8` in a separate PR since the current one already introduces a few changes (mostly related to formatting).

I agree that we can completely get rid of the `flake8` dependency, but that would require changing the logic in tests, which I would rather do in a separate PR to ensure that we don't miss anything.

**Pull request merged**